testing/src/test/java/io/stargate/it/MultipleStargateInstancesTest.java

```diff
@@ -0,0 +1,144 @@
+
+  public MultipleStargateInstancesTest(ClusterConnectionInfo backend) {
+    super(backend);
+  }
+
+  @BeforeEach
+  public void setup(TestInfo testInfo) {
+    DriverConfigLoader loader =
+        DriverConfigLoader.programmaticBuilder()
+            .withBoolean(DefaultDriverOption.METADATA_TOKEN_MAP_ENABLED, false)
+            .withString(
+                DefaultDriverOption.LOAD_BALANCING_POLICY_CLASS,
+                DcInferringLoadBalancingPolicy.class.getName())
+            .withDuration(DefaultDriverOption.REQUEST_TIMEOUT, Duration.ofSeconds(5))
+            .withDuration(DefaultDriverOption.CONNECTION_INIT_QUERY_TIMEOUT, Duration.ofSeconds(5))
+            .withDuration(DefaultDriverOption.CONTROL_CONNECTION_TIMEOUT, Duration.ofSeconds(5))
+            .withDuration(DefaultDriverOption.REQUEST_TRACE_INTERVAL, Duration.ofSeconds(1))
+            .withStringList(
+                DefaultDriverOption.METRICS_NODE_ENABLED,
+                Collections.singletonList(DefaultNodeMetric.CQL_MESSAGES.getPath()))
+            .build();
+
+    CqlSessionBuilder cqlSessionBuilder = CqlSession.builder().withConfigLoader(loader);
+    for (String host : getStargateHosts()) {
+      cqlSessionBuilder.addContactPoint(new InetSocketAddress(host, 9043));
+    }
+    session = cqlSessionBuilder.build();
+
+    Optional<String> name = testInfo.getTestMethod().map(Method::getName);
+    assertThat(name).isPresent();
+    String testName = name.get();
+    keyspace = "ks_" + testName;
+    table = testName;
+    runningStargateNodes = stargateStarters.size();
+  }
+
+  @Test
+  public void shouldConnectToMultipleStargateNodes() {
+    List<Row> all = session.execute("SELECT * FROM system.peers").all();
+    // system.peers should have N records (all stargate nodes - 1)
+    assertThat(all.size()).isEqualTo(runningStargateNodes - 1);
+  }
+
+  @Test
+  public void shouldDistributeTrafficUniformly() {
+    // given
+    createKeyspaceAndTable();
+    long totalNumberOfRequests = 300;
+    long numberOfRequestPerNode = totalNumberOfRequests / runningStargateNodes;
+    // difference tolerance - every node should have numberOfRequestPerNode +- tolerance
+    long tolerance = 5;
+
+    // when
+    for (int i = 0; i < totalNumberOfRequests; i++) {
+      session.execute(
+          SimpleStatement.newInstance(
+              String.format(
+                  "INSERT INTO \"%s\".\"%s\" (k, cc, v) VALUES (1, ?, ?)", keyspace, table),
+              i,
+              i));
+    }
+
+    // then
+    Collection<Node> nodes = session.getMetadata().getNodes().values();
+    assertThat(nodes.size()).isEqualTo(runningStargateNodes);
+    for (Node n : nodes) {
+      long cqlMessages =
+          ((Timer)
+                  session.getMetrics().get().getNodeMetric(n, DefaultNodeMetric.CQL_MESSAGES).get())
+              .getCount();
+      assertThat(cqlMessages)
+          .isBetween(numberOfRequestPerNode - tolerance, numberOfRequestPerNode + tolerance);
```

Comment:
Where does the variance come from?... just wondering :thinking: shouldn't the driver do strict round-robin?

Reply:
I didn't analyze it more deeply, in a variety of cases it is equal to the exact number but I observed some cases when it was not even. It can be for example retry or speculative execution.